src/main/java/fi/jubic/quanta/models/QuantaAuthenticator.java

```diff
@@ -0,0 +1,73 @@
+package fi.jubic.quanta.models;
+
+import fi.jubic.quanta.dao.ExternalClientDao;
+import fi.jubic.snoozy.auth.implementation.StatefulAuthenticator;
+import fi.jubic.snoozy.auth.implementation.Token;
+
+import javax.inject.Inject;
+import javax.inject.Singleton;
+import java.time.LocalDateTime;
+import java.util.List;
+import java.util.Optional;
+import java.util.concurrent.CopyOnWriteArrayList;
+import java.util.stream.Collectors;
+
+@Singleton
+public class QuantaAuthenticator extends StatefulAuthenticator<User> {
+    private List<Token<User>> tokens;
+    private final ExternalClientDao externalClientDao;
+
+    @Inject
+    public QuantaAuthenticator(
+            ExternalClientDao externalClientDao
+    ) {
+        this.tokens = new CopyOnWriteArrayList<>();
+        this.externalClientDao = externalClientDao;
+        reloadExternalClients();
+    }
+
+    @Override
+    public Optional<User> authenticate(String token) {
+        if (super.authenticate(token).isPresent()) {
+            return super.authenticate(token);
+        }
+        return tokens.stream()
+                .filter(t -> t.getToken().equals(token))
+                .findFirst()
+                .map(Token::getUser);
+    }
+
+    public void reloadExternalClients(ExternalClient externalClient) {
+        Token<User> newToken = new Token<>(
+                externalClient.getUser(),
+                externalClient.getToken(),
+                LocalDateTime.now().plusYears(100)
+        );
+        if (!tokens.contains(newToken)) {
+            tokens.add(newToken);
+            authenticate(newToken.getToken());
+        }
+    }
+
+    public List<Token<User>> reloadExternalClients() {
```

Comment:
This method should return `void`. The result does not seem to be even used.

Reply:
Changed to `void`